feat: update Shadow study extension to v1.1.0

j-russo opened the pull request:
# Shadow Study Extension Update to v1.1.0
## Overview
This PR enhances the Shadow Study extension with several new features focused on improving user workflow, adding multiple date support, and providing more export options including animated GIFs.

### New Features

**1. Multiple Date Support**
 * Add multiple dates
 * Preview any selected date from multiple dates
 
**2. Specific Time Support**
 * Direct selection of specific moment in time

**3. Export Updates**
 * Export multiple dates at once
    * Organized folder structure by date
 * Export time range as animated GIF
 * Export specific moment in time as PNG

**4. UI Updates**
 * Organized into sections
 * Default date set to June 21 (summer solstice)
## Technical Notes
* Addressed viewport flashing during batch exports in collaboration with Forma team (added loading spinner), which helps prevent potential photosensitivity issues
* Maintained compatibility with existing extension
* Added TODOs for potential future enhancements
## Testing
The extension has been tested with:
* Various date combinations
* Different time ranges and intervals
* Multiple export formats (PNG, ZIP, GIF)
* Different screen resolutions
## Screenshots

UI - Time Range
![250430_shadow_study_1_1_0_ui_range](https://github.com/user-attachments/assets/7864f466-7000-49f0-b6df-5bd4a87cd3a6)

UI - Specific Time
![250430_shadow_study_1_1_0_ui_specific](https://github.com/user-attachments/assets/870764b2-b5b2-464f-83de-59f8ab590893)

Export Flashing Fix - Loading Spinner
![250430_shadow_study_1_1_0_export_loading_spinner](https://github.com/user-attachments/assets/38e9da6e-e1ff-4458-b17b-959eb519c86e)


Thank you for reviewing this update. I'm happy to address any questions or feedback.

j-russo committed:
feat: update Shadow study extension to v1.1.0

j-russo committed:
fix: correct export button text in specific time mode

- Add useEffect to reset isGif when switching to specific time mode
- Update export button text to show "Export image" (singular) in specific time mode
- Maintain "Export images" or "Export animation" text in range mode

This ensures the button text accurately reflects what will be exported.

henrik-steen-adsk committed:
Merge branch 'main' into feat/shadow-study-v1.1.0

henrik-steen-adsk commented on package.json:
```diff
@@ -1,18 +1,20 @@
   "name": "shadow-study",
   "private": true,
-  "version": "1.0.0",
+  "version": "1.1.0",
```
We don't maintain any version specifier here, so keep this as 1.0.0 (it can really be whatever, as it's not really used for anything.)
```suggestion
  "version": "1.0.0",
```

j-russo replied:
Kept version as 1.0.0.

henrik-steen-adsk commented on index.html:
```diff
@@ -18,7 +18,7 @@
       src="https://app.autodeskforma.eu/design-system/v2/weave/components/checkbox/weave-checkbox.js"
     ></script>
-    <title>Shadow study</title>
+    <title>Shadow study (SOM)</title>
```
What does SOM mean?

Btw this title isn't shown anywhere since extensions load as iframes.

j-russo replied:
Removed "(SOM)" from the title. (SOM is my company's name, I was using it to differentiate from the original during testing.)

henrik-steen-adsk commented on forma-shadow-study-questions.md:
This file should be removed before this PR can be merged.

j-russo replied:
Removed this file.

henrik-steen-adsk commented on forma-shadow-study-questions.md:
```diff
@@ -0,0 +1,78 @@
+A more robust solution required Forma API enhancements to support "silent" captures without viewport updates, which the Forma team has now implemented.
+
+### 2. Extension Icons and Manifest Implementation
+Based on reviewing the Forma publishing documentation and original repository (which I noticed doesn't include a manifest.json), it looks like the extension icons and manifest.json might be handled during the publishing process rather than being required during development. I tried several approaches to get the icons to display during development but was unsuccessful.
```
What documentation have you come across that mention `manifest.json`? We don't use this file for anything. Icons are configured as part of the extension configuration in the Forma UI.

j-russo replied:
That was an initial question I had but was clarified by the team as well. Thank you for confirming.

henrik-steen-adsk commented on manifest.json:
This file has no purpose and should be removed

j-russo replied:
Removed.

henrik-steen-adsk commented on package-lock.json:
This repo currently uses yarn instead of of NPM. I see the `yarn.lock` file has been updated as well, so I guess both `yarn` and `npm` commands have been used. Remove this file.

j-russo replied:
Removed package-lock.json to use only yarn and yarn.lock for dependency management.

henrik-steen-adsk commented on package.json:
```diff
@@ -1,18 +1,20 @@
     "prettier:check": "yarn prettier --check src",
-    "prettier:write": "yarn prettier --write src"
+    "prettier:write": "yarn prettier --write src",
+    "dev": "vite"
```
What is the purpose of this run-script? `start` should already cover this

j-russo replied:
Cleaned up scripts.

henrik-steen-adsk commented on package.json:
```diff
@@ -1,18 +1,20 @@
   "type": "module",
   "scripts": {
-    "start": "DEV_SERVER=1 vite",
+    "start": "cross-env DEV_SERVER=1 vite",
```
I see `DEV_SERVER=1` is unused in this repo and something from other internal repos. You can also remove the `cross-env` dependency.

```suggestion
    "start": "vite",
```

j-russo replied:
Removed unused DEV_SERVER and cross-env.

henrik-steen-adsk commented on public/gif.worker.js:
Don't vendor external code. Usually needs like this can be solved by vite and special import specifiers. (But see comments on gif.js not being maintained before digging into this.)

j-russo replied:
Removed vendored gif.worker.js and migrated GIF export to use modern-gif, which is actively maintained and managed as a dependency.

henrik-steen-adsk commented on package.json:
```diff
@@ -1,18 +1,20 @@
   "dependencies": {
     "file-saver": "^2.0.5",
     "forma-embedded-view-sdk": "^0.87.0",
+    "gif.js": "^0.2.0",
```
`gif.js` does not seem to be maintained as it was lasted updated 6 years ago. Please see if you can find another dependency that can solve the need that is maintained.

j-russo replied:
Replaced gif.js with modern-gif, a maintained and modern alternative for GIF encoding.

henrik-steen-adsk commented on src/app.tsx:
```diff
@@ -1,27 +1,96 @@
+  const [year, setYear] = useState<number>(new Date().getFullYear());
+  const [isGif, setIsGif] = useState(false);
+
+  // TODO: Future enhancement - Add quick selection of standard solar analysis dates:
```
Let's keep suggestions for future functionality out of the code as it clutters the code. Feel free to create GitHub issues if you think we should keep track of it.

j-russo replied:
Removed TODO comments for future enhancements to keep code clean.

henrik-steen-adsk commented on src/components/DateSelector.tsx:
```diff
@@ -1,53 +1,80 @@
-            <weave-select-option value={value + 1}>{name}</weave-select-option>
+          {Array.from({ length: 12 }, (_, i) => i + 1).map((m) => (
+            <weave-select-option key={m} value={m}>
+              {new Date(year, m - 1).toLocaleString('default', { month: 'long' })}
```
```suggestion
              {new Date(year, m - 1).toLocaleString(undefined, { month: 'long' })}
```

j-russo replied:
Switched from 'default' to undefined in toLocaleString to ensure the user's actual locale is used for month names, improving localization.

henrik-steen-adsk commented on src/components/DateSelector.tsx:
```diff
@@ -1,53 +1,80 @@
+        <weave-button 
+          variant="outlined" 
+          onClick={() => onAddDate(month, day)}
+          data-disabled={dates.some((d) => d.month === month && d.day === day) ? "" : undefined}
```
```suggestion
          disabled={dates.some((d) => d.month === month && d.day === day)}
```

j-russo replied:
Used the disabled attribute for the Add Date button, applying a type assertion workaround due to current custom element typings not recognizing disabled. This ensures accessibility and proper button behavior while maintaining type safety.

henrik-steen-adsk replied:
We don't specify exhaustive types for attributes of web components (as we can't really guarantee that anyways), so you can you simply update the `weave.d.ts` file and add the missing declaration for `disabled`.

j-russo replied:
Thanks for the clarification! I’ve updated the local type definitions to include the `disabled` attribute for weave-button.

henrik-steen-adsk commented on src/components/PreviewButton.tsx:
```diff
@@ -1,69 +1,136 @@
+          <Fragment>
+            {dates.map((date, index) => (
+              <weave-select-option key={index} value={index}>
+                {`${MONTHS[date.month - 1]} ${date.day}`}
```
For consistency either always use MONTHS or `toLocaleString`

j-russo replied:
Standardized on toLocaleString for consistency and localization.

henrik-steen-adsk commented on src/components/TimeSelector.tsx:
```diff
@@ -11,70 +15,133 @@ type TimeSelectorProps = {
+              onChange={(event) => setStartHour(parseInt((event as CustomEvent).detail.value, 10))}
+              style={{ width: "70px" }}
+            >
+              {_.range(0, 24).map((value) => (
```
For consistency I suggest either always use lodash or `Array.from`

j-russo replied:
Standardized on lodash’s .range.

henrik-steen-adsk commented on vite.config.ts:
```diff
@@ -6,6 +6,10 @@ export default defineConfig({
   },
+  build: {
+    copyPublicDir: true
+  }
```
Was there a reason for the port change?

The hmr and copyPublicDir already have these values as default, so not sure why they where added.

j-russo replied:
Set port to 8081 and removed unnecessary default config options.

henrik-steen-adsk commented on assets/Screenshot.png:
This file is referenced from README which is now broken

j-russo replied:
Added the screenshot file back to the assets folder.

henrik-steen-adsk replied:
I think the screenshot could need an update, but I recommend doing this later before we merge to get the most up-to-date version.

henrik-steen-adsk reviewed (changes requested):
Hi @j-russo!

Thank you for your contribution! Apologize for the long time to come back to you. It took some time to get the repo set up to accept external contributions. Please see the comment above regarding the CLA.

Please also see comments I've added on the changes.

Also make sure the checks go through without errors (i.e. prettier).

j-russo committed:
Address Forma team feedback: cleanup, dependency updates, and consistency

- Removed unused files (manifest.json, package-lock.json, forma-shadow-study-questions.md, gif.worker.js)
- Migrated GIF export from gif.js to modern-gif for maintenance and performance
- Standardized on lodash _.range and toLocaleString for consistency and localization
- Cleaned up scripts, removed cross-env and DEV_SERVER, and set port to 8081
- Removed TODO comments from code
- Updated Add Date button to use disabled attribute with type workaround
- Restored screenshot in README
- Ensured all code passes formatting and linting checks

henrik-steen-adsk commented on package.json:
```diff
@@ -16,13 +16,15 @@
     "@types/file-saver": "^2.0.7",
     "@types/lodash": "4.17.0",
     "@types/luxon": "^3.4.2",
+    "cross-env": "^7.0.3",
```
This is not used and should be removed.

j-russo replied:
Removed.

j-russo committed:
fix: update weave.d.ts to support disabled attribute for weave-button

j-russo committed:
fix: format codebase with Prettier to pass CI check
